### PR TITLE
Add callback to Watching.prototype.invalidate

### LINKED ISSUE #1785: Watching.prototype.invalidate should have a callback
`Watching.prototype.invalidate` allows to recompile however it is really hard to get feedback once it is complete

https://github.com/webpack/webpack/blob/master/lib/Compiler.js#L113

Could you please add a callback?

### PR DESCRIPTION
**Pull request by kutyel**
**Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

`Watching.prototype.invalidate` allows to recompile however it is really hard to get feedback once it is complete.

**What is the new behavior?**

Now after the invalidate function is done calls a callback, and thus let's you get feedback.

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

**Other information**

Fixes #1785

Happy Hacktoberfest!! 😉

### CODE REVIEW

File: lib/Compiler.js [hunk 1]
```diff
@@ -121,6 +123,7 @@ Watching.prototype.invalidate = function() {
 	} else {
 		this._go();
 	}
+	callback();
```
Comment (selvagsz) [inline lib/Compiler.js]: @kutyel `callback` will not be invoked after the `return` statement https://github.com/webpack/webpack/pull/3103/files#diff-7a14ca420cc4fa7cf55f788246117e7cR122

**Comment by kutyel** [comment]
@selvagsz you are right, changed it! 😉

**Comment by sokra** [comment]
Your code triggers the callback once the invalidation is trigger, but instead it should be called when the compilation has finished.

**Comment by kutyel** [comment]
@sokra thanks fir the hint, any further advice on how to achieve it? Could it be easily changed within that same function definition or should I use some kind of defer method to wait for the compilation to finish? I really don't know the api, thanks in advance!

**Comment by TheLarkInn** [comment]
Take a look at the Compilation class to see where it finalizes and in Compiler.

**Comment by kutyel** [comment]
@TheLarkInn Is this any better?

**Comment by sokra** [comment]
a bit.

Best push the `callback` into an array (`this.callbacks`?) and invoke them in `this._done` and clear the array (`this.callbacks.length = 0`).

**Comment by kutyel** [comment]
Thanks @sokra 😄 I hope this one is closer 😅

File: lib/Compiler.js [hunk 2]
```diff
@@ -110,7 +110,17 @@ Watching.prototype.watch = function(files, dirs, missing) {
 	}.bind(this));
 };
 
-Watching.prototype.invalidate = function() {
+Watching.prototype.invalidate = function(callback) {
+	callback = callback || function() {};
+	this.callbacks = this.callbacks || [];
+	this.callbacks.push(callback);
+	this._done = function() {
```
Comment (sokra) [inline lib/Compiler.js]: don't *override*  the `_done` function. Just add this into the `_done` function.

Comment (kutyel) [inline lib/Compiler.js]: Like so?

**Comment by kutyel** [comment]
@sokra done :wink:

**Comment by kutyel** [comment]
I see the checks are passing except for the coverage, @sokra is the change more or less what you expected? If so, I will cover it with unit tests and that should be it, otherwise I'll change anything that you tell me and I'll try it again 😉

File: lib/Compiler.js [hunk 3]
```diff
@@ -83,6 +83,7 @@ Watching.prototype._done = function(err, compilation) {
 	this.running = false;
 	if(this.invalid) return this._go();
 	this.error = err || null;
+	this.callbacks = this.callbacks || [];
```
Comment (sokra) [inline lib/Compiler.js]: Could you just set `this.callbacks = []` in the constructor?

File: lib/Compiler.js [hunk 4]
```diff
@@ -110,7 +115,11 @@ Watching.prototype.watch = function(files, dirs, missing) {
 	}.bind(this));
 };
 
-Watching.prototype.invalidate = function() {
+Watching.prototype.invalidate = function(callback) {
+	callback = callback || function() {};
+	this.callbacks = this.callbacks || [];
```
Comment (sokra) [inline lib/Compiler.js]: same

File: lib/Compiler.js [hunk 5]
```diff
@@ -110,7 +115,11 @@ Watching.prototype.watch = function(files, dirs, missing) {
 	}.bind(this));
 };
 
-Watching.prototype.invalidate = function() {
+Watching.prototype.invalidate = function(callback) {
+	callback = callback || function() {};
```
Comment (sokra) [inline lib/Compiler.js]: Instead of this, just do this:

``` js
if(callback)
  this.callbacks.push(callback);
```

**Comment by kutyel** [comment]
Done, thanks for your feedback @sokra 😉

**Comment by kutyel** [comment]
@sokra if there is anything else I can help with getting this merged, just tell me 😉

**Comment by kutyel** [comment]
I just merged from master and change from es5 to es6 as I saw the codebase changed as well, I think this is ready for merge @sokra :wink: :+1:

**Pull request merged**